### PR TITLE
WIP: ElectronDresser

### PR DESCRIPTION
**Pull request by andb0t**
This package dresses electrons with close-by photons. It is currently under development and this PR serves as baseline for discussion.

BEGINRELEASENOTES
- this package dresses leptons with close-by particles. By default it dresses electrons with photons.

ENDRELEASENOTES

### CODE REVIEW

File: Analysis/ElectronDresser/src/ElectronDresser.cc [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+  for (int i = 0; i < _lep_n; ++i) {
+    _fourvec.SetPtEtaPhiE(_lep_pt.at(i), _lep_eta.at(i), _lep_phi.at(i), _lep_e.at(i));
+    for (int j = 0; j < _pfo_n; ++j) {
+      if (_pfo_type.at(j) == 22) {
+        _tmpvec.SetPtEtaPhiE(_pfo_pt.at(j), _pfo_eta.at(j), _pfo_phi.at(j), _pfo_e.at(j));
+        double dR = DeltaR(_fourvec.Eta(), _fourvec.Phi(), _tmpvec.Eta(), _tmpvec.Phi());
+        printf("Analyzing lepton %d and photon pfo %d: dR = %f\n", i, j, dR);
+        if (dR < ELECTRON_DRESS_MAX_DR) {
+          if (std::find(usedPFOs.begin(), usedPFOs.end(), j) != usedPFOs.end()) {
+            printf(
```
Comment (rete) [inline Analysis/ElectronDresser/src/ElectronDresser.cc]: Can use streamlog_out instead printf please ?

File: Analysis/ElectronDresser/src/ElectronDresser.cc [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+ElectronDresser aElectronDresser;
+
+ElectronDresser::ElectronDresser() : Processor("ElectronDresser") {
+  // modify processor description
+  _description = "ElectronDresser does whatever it does ...";
+
+  // register parameters: collection type, aribtrary name, arbitrary description, class-variable, default value
+  // input
+  registerInputCollection(LCIO::RECONSTRUCTEDPARTICLE, "", "", _m_pfos, std::string("LooseSelectedPandoraPFANewPFOs"));
```
Comment (rete) [inline Analysis/ElectronDresser/src/ElectronDresser.cc]: Can you add a description to your parameters please ? This is used by Marlin -x to document processor parameters.

File: Analysis/ElectronDresser/src/ElectronDresser.cc [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        double dR = DeltaR(_fourvec.Eta(), _fourvec.Phi(), _tmpvec.Eta(), _tmpvec.Phi());
+        printf("Analyzing lepton %d and photon pfo %d: dR = %f\n", i, j, dR);
+        if (dR < ELECTRON_DRESS_MAX_DR) {
+          if (std::find(usedPFOs.begin(), usedPFOs.end(), j) != usedPFOs.end()) {
+            printf(
+                "Warning: two leptons to be dressed with same PFO! Should not happen if dR_leptons > dR_dressRadius! Skip "
+                "this PFO!\n");
+            continue;
+          }
+          printf("Got one! Dressing Lepton %d with PFO %d: dR = %f, type %d\n", i, j, dR, _pfo_type.at(j));
```
Comment (rete) [inline Analysis/ElectronDresser/src/ElectronDresser.cc]: Same here

File: Analysis/ElectronDresser/src/ElectronDresser.cc [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    dressedCollection->addElement(pReconstructedParticle);
+  }
+
+  evt->addCollection(dressedCollection, _m_DressedLepton);
+}
+void ElectronDresser::getCollection(LCCollection*& collection, std::string collectionName, LCEvent* evt) {
+  try {
+    collection = evt->getCollection(collectionName);
+  } catch (DataNotAvailableException& e) {
+    std::cout << "- cannot get collections !!" << std::endl;
```
Comment (rete) [inline Analysis/ElectronDresser/src/ElectronDresser.cc]: Same here

**Comment by rete** [comment]
Please also use the tags for your PR as mentioned here : https://github.com/iLCSoft/MarlinReco/blob/master/.github/PULL_REQUEST_TEMPLATE

File: Analysis/ElectronDresser/src/ElectronDresser.cc (hunk 2 above)
Comment (andb0t) [reply Analysis/ElectronDresser/src/ElectronDresser.cc]: Done!

File: Analysis/ElectronDresser/src/ElectronDresser.cc (hunk 1 above)
Comment (andb0t) [reply Analysis/ElectronDresser/src/ElectronDresser.cc]: Done!

File: Analysis/ElectronDresser/src/ElectronDresser.cc (hunk 3 above)
Comment (andb0t) [reply Analysis/ElectronDresser/src/ElectronDresser.cc]: Done!

File: Analysis/ElectronDresser/src/ElectronDresser.cc (hunk 4 above)
Comment (andb0t) [reply Analysis/ElectronDresser/src/ElectronDresser.cc]: Done!

**Comment by andb0t** [comment]
The electron dressing has to be implemented within the IsolatedLeptonFinder. I will do so and submit another PR once it is ready.

**Pull request closed without merging**